Extend #619: Seed creation with coin flips

Stargazer10101 opened the pull request:
## Description

Builds on #619 .
This PR aims to complete the development of the "Create Seed with Coin" tool.

New screens added as a part of this PR:
<img width="240" height="240" alt="ToolsCoinEntropyMnemonicLengthView" src="https://github.com/user-attachments/assets/747ab5a3-742b-4977-9a12-bfbe400588f1" />
<img width="240" height="240" alt="ToolsCoinInputMethodView_12words" src="https://github.com/user-attachments/assets/7b9561e4-9af9-495d-b4e2-73442f029373" />
<img width="240" height="240" alt="ToolsCoinEntropyEntryView_12words" src="https://github.com/user-attachments/assets/d15122ed-c890-4dc6-bec6-c2b2637acca1" />
<img width="240" height="240" alt="ToolsCoinEntropySetwiseEntryView_12words_set1" src="https://github.com/user-attachments/assets/80653bb2-a371-4dae-ad98-5af63a918d4f" />
<img width="240" height="240" alt="ToolsCoinEntropySetwiseBip39WordView_12words" src="https://github.com/user-attachments/assets/32f8f8e7-0b39-4ae0-9b67-40b1a0deea70" />



Related: https://github.com/SeedSigner/seedsigner/pull/619
This pull request is categorized as a:

- [x] New feature
- [ ] Bug fix
- [ ] Code refactor
- [ ] Documentation
- [ ] Other

## Checklist

- [x] I’ve run `pytest` and made sure all unit tests pass before sumbitting the PR
![Screenshot from 2025-04-09 20-43-38](https://github.com/user-attachments/assets/06bbd130-ea9c-41c1-bd61-aefac04c6f17)


If you modified or added functionality/workflow, did you add new unit tests?

- [ ] No, I’m a fool
- [x] Yes
- [ ] N/A

I have tested this PR on the following platforms/os:

- [x] Raspberry Pi OS [Manual Build](https://github.com/SeedSigner/seedsigner/blob/dev/docs/manual_installation.md)
- [ ] [SeedSigner OS](https://github.com/SeedSigner/seedsigner-os) on a Pi0/Pi0W board
- [ ] Other

BtcAutoNode committed:
Update tools_views.py

- add menu entry and views

BtcAutoNode committed:
Update components.py

- add icons (coins ion for the seed menu and T/H for head and tails)

BtcAutoNode committed:
Update tools_screens.py

- add coin flip screen

BtcAutoNode committed:
Update mnemonic_generation.py

- add number of flips constants

BtcAutoNode committed:
Update tools_screens.py

- change layout to 3, 3 similar to the dice screen

BtcAutoNode committed:
Update tools_screens.py

- change order of H and T

BtcAutoNode committed:
Update tools_views.py

- add missing space in menu entry

BtcAutoNode committed:
Update test_mnemonic_generation.py

- add pytest tests similar to the tests for dice rolls

BtcAutoNode committed:
Update test_mnemonic_generation.py

- fix copy/paste error

BtcAutoNode committed:
Update components.py

- delete heads/tails icon (and stick to normal letters instead of icons (as easyuxd proposed)

BtcAutoNode committed:
Update tools_screens.py

- changed coin flip screen) to reflect changes proposed by easyuxd)

Stargazer10101 committed:
feat: Add input string length and contents verification in mnemonic_generation

kdmukai commented:
Will need to be updated to use the new l10n / multilanguage patterns. Review the doc in the new `l10n` folder.

"H" and "T" will also have to be carefully wrapped for translation!

But overall this will be a nice additional feature for completeness.

Stargazer10101 committed:
chore: Resolve Conflict in tools_views.py

Advaitgaur004 commented:
![Screenshot from 2025-04-10 13-26-33](https://github.com/user-attachments/assets/cac5afea-b59e-453a-a556-8a60fec6dbb1)
@Stargazer10101 just tested on this PR , Exception Raised .
```python 
raise Exception("Refactor to ButtonOption approach needed!")
```
in `ButtonListScreen` . Does this works on Emulator. If Yes, how to reproduce this

Stargazer10101 committed:
update: tools_views.py to use l10n patterns

Stargazer10101 committed:
update: tools_screen.py to use l10n patterns

Stargazer10101 committed:
fix: fixed type error in tools_screens.py

kdmukai commented on src/seedsigner/views/tools_views.py:
```diff
@@ -255,12 +255,12 @@ def run(self):
-        TWELVE = f"12 words ({mnemonic_generation.COIN__NUM_FLIPS__12WORD} flips)"
-        TWENTY_FOUR = f"24 words ({mnemonic_generation.COIN__NUM_FLIPS__24WORD} flips)"
+        TWELVE = ButtonOption(f"12 words ({mnemonic_generation.COIN__NUM_FLIPS__12WORD} flips)")
+        TWENTY_FOUR = ButtonOption(f"24 words ({mnemonic_generation.COIN__NUM_FLIPS__24WORD} flips)")
```
Named variables in translation strings have to be translator-friendly.

If the python source is:
```python
greeting = _("Hello, {first_name}!")
```

the translators will see: "Hello, {first_name}!".

A spanish translator would then type: "¡Hola, {first_name}!"

Easy.

But if the python instead is:
```python
greeting = _("Hello, {self.invoices[cur_invoice.invoice_num].buyer.user.first_name}!")
```

Then the translator must also carefully match all the pythonic referencing exactly:

"¡Hola, {self.invoices[cur_invoice.invoice_num].buyer.user.first_name}!"

Which is obviously a total disaster since most translators will have no programming experience. And actually I'm not even sure if any python in there will work... Probably ONLY a simple named variable can be inserted...?

Either way, instead we'd want:
```python
first_name = self.invoices[cur_invoice.invoice_num].buyer.user.first_name
greeting = _("Hello, {first_name}!")
```

LONG way of saying: don't use the constants as the named variables.

kdmukai commented on src/seedsigner/gui/screens/tools_screens.py:
```diff
@@ -175,7 +175,7 @@ def update_title(self) -> bool:
     def __post_init__(self):
         # Override values set by the parent class
-        self.title = f"Coin Flip 1/{self.return_after_n_chars}"
+        self.title = _(f"Coin Flip 1/{self.return_after_n_chars}")
```
Same concern here. Just avoid using f-strings for strings that need translations. In this case you can refactor to:

```python
# TRANSLATOR_NOTE: Inserts the number of flips required
self.title = _("Coin Flip 1/{}").format(self.return_after_n_chars)
```

kdmukai commented on src/seedsigner/gui/screens/tools_screens.py:
```diff
@@ -180,7 +180,7 @@ def __post_init__(self):
         self.rows = 2
         self.cols = 2
-        self.key_height = GUIConstants.TOP_NAV_TITLE_FONT_SIZE + 2 + 2*GUIConstants.EDGE_PADDING
+        self.key_height = GUIConstants.get_top_nav_title_font_size() + 2 + 2*GUIConstants.EDGE_PADDING
```
We're not totally rigorous about this everywhere, but ideally there shouldn't be any hard-coded pixel values (the +2 after the title font size).

Whenever possible, stick to scaling a GUI constant (e.g. `int(GUIConstants.COMPONENT_PADDING / 4)`)

Stargazer10101 committed:
chore: address PR feedback

Stargazer10101 commented:
Hey @kdmukai . Thanks for the feedback! I have addressed all the issues pointed out. Apart from that, I've created a flow test and I'll be writing more.

Stargazer10101 committed:
Update tools_views.py

Stargazer10101 committed:
feat: enter entropy in sets of 11 coin flips

Stargazer10101 committed:
feat: add checksum generation and get_bip39_word functions to  mnemonic_generation.py

kdmukai commented on src/seedsigner/gui/screens/tools_screens.py:
```diff
@@ -171,6 +171,73 @@ def update_title(self) -> bool:
+        # Specify the keys in the keyboard
+        self.rows = 2
+        self.cols = 2
+        self.key_height = GUIConstants.get_top_nav_title_font_size() + GUIConstants.EDGE_PADDING/4 + 2*GUIConstants.EDGE_PADDING
```
Always force any division for pixel-related placement to `int`. There's no guarantee that different pixel-level constants will always be even numbers.

kdmukai commented on src/seedsigner/gui/screens/tools_screens.py:
```diff
@@ -171,6 +171,73 @@ def update_title(self) -> bool:
+        ))
+
+    def update_title(self) -> bool:
+        self.title = _("Coin Flip {}/{}").format(self.cursor_position + 1, self.return_after_n_chars )
```
Always provide a `# TRANSLATOR_NOTE` to explain the expected values for any variable insertions.

And when there's more than one variable, it's probably better to use a named variable:
```python
self.title = _(f"Coin Flip {cur_flip_num}/{num_total_flips}")
```

kdmukai commented on src/seedsigner/gui/screens/tools_screens.py:
```diff
@@ -171,6 +171,73 @@ def update_title(self) -> bool:
+    required_bits: int = 11
+
+    def __post_init__(self):
+        self.title = _(f"Set {self.current_set}/{self.total_sets}")
```
see: https://github.com/SeedSigner/seedsigner/pull/720#discussion_r2037585915

tldr: we wouldn't want translators to have to write `self.current_set`; they should never have to worry about python syntax other than the variable brackets / named variables (which is already less than ideal but can't be avoided).

kdmukai replied:
Also I wouldn't use the term "Set" in the title. The user is flipping coins to create a word in their mnemonic so "Word" should be used.

kdmukai commented on src/seedsigner/gui/screens/tools_screens.py:
```diff
@@ -171,6 +171,73 @@ def update_title(self) -> bool:
+        super().__post_init__()
+
+        self.components.append(TextArea(
+            text=_("(H)eads=1  (T)ails=0"),
```
Always need a `# TRANSLATOR_NOTE` for any string that has any potential to be confusing; translators see these strings completely out of context and out of order. Yes, it should be mostly obvious, but imagine a translator who is unfamiliar with the English heads/tails terminology.

kdmukai commented on src/seedsigner/helpers/mnemonic_generation.py:
```diff
@@ -121,3 +131,17 @@ def generate_mnemonic_from_image(image, wordlist_language_code: str = SettingsCo
     return bip39.mnemonic_from_bytes(hash.digest(), wordlist=Seed.get_wordlist(wordlist_language_code)).split()
+
+
+def generate_checksum_from_entropy(entropy_bits: str) -> str:
```
This is unnecessary. `embit`'s `mnemonic_from_bytes` will calculate it for us.

see my proposed implementation: https://gist.github.com/kdmukai/1a666c945c46ada120fd2cd3c5a1f33c

kdmukai commented on src/seedsigner/views/tools_views.py:
```diff
@@ -241,6 +250,170 @@ def run(self):
+
+
+
+class ToolsCoin_Input_Method_View(View):
```
Does not conform to our `View` naming conventions.

kdmukai commented on src/seedsigner/views/tools_views.py:
```diff
@@ -241,6 +250,170 @@ def run(self):
+
+
+
+class ToolsCoinEntropyEntryView(View):
```
Since this implementation offers two options, it is always helpful to re-explain which option this `View` is for in a comment.

kdmukai commented on src/seedsigner/views/tools_views.py:
```diff
@@ -241,6 +250,170 @@ def run(self):
+
+
+
+class ToolsCoinEntropySetwiseEntryView(View):
```
It's more obvious from the name, but still a good idea to have an explanatory comment for the `View`.

kdmukai commented on src/seedsigner/views/tools_views.py:
```diff
@@ -241,6 +250,170 @@ def run(self):
+            word_index = int(ret, 2)
+            word = Seed.get_wordlist(self.settings.get_value(SettingsConstants.SETTING__WORDLIST_LANGUAGE))[word_index]
+            self.mnemonic.append(word)
+            ButtonListScreen(
```
For the sake of simplicity a `View` should only ever invoke one `Screen` class (there is perhaps ONE exception to this rule elsewhere in the code, but otherwise it is considered a CRUCIAL rule).

Why?
* Follow-up `Screen` calls cannot be invoked by the screenshot generator so we would be missing whole portions of the user flow in our screenshots.
* Calling multiple `Screen`s / performing multiple stages of user interaction makes it MUCH harder to debug, maintain, and just reason about in general. If you look back at the code before v0.5.0, you can see the trouble that's caused by implementing multi-step flows as a single giant function. 
* tldr: A `View` should present one user interaction (one `Screen`) and then pass off execution to the next `View`.
* Provides reusability if some other user flow can leverage the same `View`/`Screen`.


How to implement: see `seed_views.SeedMnemonicEntryView`. It provides an exact parallel to the set-wise coin flip entry. The only difference is that set-wise coin entry would keep repeating two sequences:
* ToolsCoinEntropySetwiseEntryView (for each word n)
* ToolsCoinEntropySetwiseBip39WordView (or something like that) (for each word n)

Stargazer10101 committed:
fix: refactor coin flips views as per PR feedback

Stargazer10101 commented:
Thanks for the review. I've made the requested changes.

Stargazer10101 committed:
fix: rename ToolsCoin_Input_Method_View to ToolsCoinInputMethodView

Stargazer10101 commented:
I've tested it on pi 0. Are there any remarks/feedback? Should I squash the changes made to the same file multiple times into  a single commit for each file?
I'll proceed to writing tests next.

Stargazer10101 commented:
Right now the screen looks like this:
![Screenshot from 2025-06-22 01-55-24](https://github.com/user-attachments/assets/9694033c-470e-46f7-8259-8eb3d9330cfa)

I think there should be a number getting displayed that tells how many coin flips have been done so far(out of the 11 required for the current word).

Stargazer10101 committed:
fix: mnemonic generation for option 1: enter entropy in one go

Stargazer10101 commented:
For the "128/256 coin flips in one go" option, the mnemonic sentence getting generated was incorrect. The mnemonic generation logic has now been corrected. Verified with iancoleman https://iancoleman.io/bip39/. And tested on raspberry pi 0.

Stargazer10101 committed:
fix: test_mnemonic_generation.py

Stargazer10101 commented:
I've updated the test cases to assert against the "Use raw entropy" option in iancoleman. Previously, the mnemonic generation logic and the test expectations were aligned with the options- "12 words" or "24 words" in the iancoleman tool. The mnemonics generated using these options use random number generators in addition with user-provided entropy. It is more suitable for the case where user isn't so sure about the randomness of their entropy and they trust the software's RNG. 

This is completely against what we are trying to achieve. The users are providing raw entropy and would want that entropy to map to the seed phrase as closely as possible, without any RNGs in the middle. The coin is the only RNG our users want to trust. Considering this, the tests have been changes accordingly.

Stargazer10101 committed:
fix: fix views navigation in create-seed-using-coin tool

notTanveer commented on src/seedsigner/views/tools_views.py:
```diff
@@ -241,6 +250,192 @@ def run(self):
+            raise ValueError("Unsupported flip count")
+
+        ALL_FLIPS = ButtonOption(all_flips_text, return_data="all")
+        SETWISE = ButtonOption(setwise_text, return_data="setwise")
```
```suggestion
        all_flip_number = 128 if self.total_flips == 128 else 256
        all_flips = _("{} coin flips in one go").format(all_flip_number)

        ALL_FLIPS = ButtonOption(all_flips, return_data="all")
        SETWISE = ButtonOption(_("Sets of 11 coin flips"), return_data="setwise")
```

notTanveer commented:
was testing and found a weird loop. when you hit the back button, it just keeps bouncing between the coin mnemonic screen and the tools view forever.

notTanveer commented on src/seedsigner/views/tools_views.py:
```diff
@@ -241,6 +250,192 @@ def run(self):
+        ).display()
+
+        if selected_menu_num == RET_CODE__BACK_BUTTON:
+            return Destination(ToolsMenuView)
```
this should fix the infinite back loop.
```suggestion
            return Destination(ToolsMenuView, clear_history=True)
```

notTanveer commented on src/seedsigner/gui/screens/tools_screens.py:
```diff
@@ -171,6 +171,92 @@ def update_title(self) -> bool:
+        # TRANSLATOR_NOTE: Inserts the number of flips required
+        self.title = _("Coin Flip 1/{}").format(self.return_after_n_chars)
+
+        # Specify the keys in the keyboard
```
the comments in this class feels kinda unnecessary from this point onward

Stargazer10101 committed:
add flow tests

Stargazer10101 committed:
fix infinite loop in UI

Stargazer10101 commented:
Thanks for the review @notTanveer . I've also added some flow tests and they're are not yet passing and I'm working on it. There's some issue related to MagicMock. I'll fix that and later I'll squash all flow-tests related commits into one commit.

Stargazer10101 commented:
I'm stuck at the MagickMock comparison issue. MagicMock objects are being used to simulate parts of the system, but the new views (like ToolsCoinInputMethodView) are trying to compare these MagicMock objects with real integers, causing TypeError: '<' not supported between instances of 'MagicMock' and 'int' errors.
<img width="898" height="299" alt="image" src="https://github.com/user-attachments/assets/6da2909c-bd9d-47d1-ab9c-af6c9fe1b8f3" />

In the commit [91a67ef](https://github.com/SeedSigner/seedsigner/pull/720/commits/91a67efb944af21bee4383d447508f05a3f55e73) I attempted to fix this issue, but now the tests are hanging (the tests don't run to completion, it just hangs).

Stargazer10101 commented:
@kdmukai can you please take a look?

kdmukai commented on src/seedsigner/views/tools_views.py:
```diff
@@ -241,6 +250,189 @@ def run(self):
+        TWENTY_FOUR = ButtonOption(twenty_four, return_data=mnemonic_generation.COIN__NUM_FLIPS__24WORD)
+
+        button_data = [TWELVE, TWENTY_FOUR]
+        selected_menu_num = ButtonListScreen(
+            title=_("Mnemonic Length"),
+            is_bottom_list=True,
+            is_button_text_centered=True,
+            button_data=button_data,
+        ).display()
```
Ah, here's the problem. You had the bad luck of basing this on older `View` code that was not yet updated for the `self.run_screen()` call that is required for the flow tests. If a View still use the old `Screen.display()` method, that means that a flow test has not yet been written for that View, which indicates a testing gap. ALL Views should be updated to the `self.run_screen()` method.

```suggestion
        selected_menu_num = self.run_screen(
            ButtonListScreen,
            title=_("Mnemonic Length"),
            is_bottom_list=True,
            is_button_text_centered=True,
            button_data=button_data,
        )
```

Stargazer10101 committed:
update the views to self.run_screen method and update flow tests

Stargazer10101 committed:
fix: resolve merge conflicts

Stargazer10101 commented:
The is_redirect=True in these lines is causing test failures. Also, the comments indicate that they might not be necessary. Should I just remove these?

<img width="1344" height="540" alt="image" src="https://github.com/user-attachments/assets/e20ff426-1668-4145-9067-8b1a0181f218" />


Here is the first instance of it:
https://github.com/SeedSigner/seedsigner/blob/59836e7bc31512eb5b8b18a6304dc25d773f6a62/tests/test_flows_seed.py#L452

Stargazer10101 committed:
Merge branch 'dev' into create_seed_with_coin

Stargazer10101 committed:
test: remove is_redirect=True from SeedTranscribeSeedQRZoomedInView flow tests

Stargazer10101 commented:
@kdmukai Review Requested

alvroble commented on src/seedsigner/gui/screens/tools_screens.py:
I suggest updating the screenshot generator to include the new screens introduced in this PR.

alvroble commented on src/seedsigner/views/seed_views.py:
My take here is that the `.display()` to `self.run_screen()` migration maybe should happen in a separate PR, I don't know what others devs think.

Chaitanya-Keyal replied:
Ran into this issue while working on my BIP-85 PR too, I've opened #801 that migrates this across the codebase.

kdmukai replied:
I agree with @alvroble. This PR should NOT include the `seed_views.py` edits, even though they absolutely SHOULD be updated, just in a separate, clean, simple PR.

Stargazer10101 committed:
add seed with coinflips views

Stargazer10101 committed:
remove is_redirect=True from SeedTranscribeSeedQRZoomedInView flow test

Stargazer10101 committed:
Fix PIL paste coordinate float issue

Stargazer10101 commented:
@kdmukai Can you please run the remaining workflows (the ones requiring maintainer approval)?

notTanveer commented:
i'll suggest to add the new screens, in the PR description.

notTanveer commented on src/seedsigner/gui/components.py:
```diff
@@ -653,7 +654,7 @@ def run(self):
 
                     img = self.rendered_text_img.crop((self.horizontal_scroll_position, 0, self.horizontal_scroll_position + self.visible_width, self.rendered_text_img.height))
-                    self.renderer.canvas.paste(img, (self.screen_x, self.screen_y - self.scroll_y))
+                    self.renderer.canvas.paste(img, (int(self.screen_x), int(self.screen_y - self.scroll_y)))
```
revert. this is already int

notTanveer commented on src/seedsigner/gui/components.py:
```diff
@@ -716,7 +717,7 @@ def render(self):
             text_img = text_img.crop((0, 0, self.visible_width, text_img.height))
 
-        self.canvas.paste(text_img, (text_x, self.screen_y + self.text_y - self.text_height_above_baseline - self.scroll_y))
+        self.canvas.paste(text_img, (int(text_x), int(self.screen_y + self.text_y - self.text_height_above_baseline - self.scroll_y)))
```
same here. already int

notTanveer commented on tests/test_flows_seed.py:
```diff
@@ -477,6 +477,60 @@ def load_right_seed_into_decoder(view: View):
         ])
 
 
+    def test_coin_flip_flow(self):
```
i think better naming would be, 
```suggestion
    def test_coin_flip_seed_flow(self):
```

Stargazer10101 committed:
fix float coordinate issue in coin entropy screens

Stargazer10101 committed:
Rename test_coin_flip_flow to test_coin_flip_seed_flow

Stargazer10101 committed:
rename the SetwiseBip39WordView title from set -> word

kdmukai commented:
> @kdmukai Can you please run the remaining workflows (the ones requiring maintainer approval)?

Note that I am NOT a maintainer of the repo!! I actually have zero privileges!

This is an intentional separation of concerns that I've insisted on to limit my control over the project.

kdmukai commented on src/seedsigner/gui/screens/tools_screens.py:
```diff
@@ -198,6 +198,92 @@ def update_title(self) -> bool:
+        self.keys_charset = "".join([
+            "H",
+            "T",
+        ])
```
The keys in the keyboard must be localized.

And after some quick googling, Spain uses "cara" / "cruz" so the UI can't rely on a single letter being sufficient.

Since the keyboard buttons are being rendered so large, I think we should see if the whole word will fit.

kdmukai commented on src/seedsigner/gui/screens/tools_screens.py:
```diff
@@ -198,6 +198,92 @@ def update_title(self) -> bool:
+        self.title = _("Word {current_set}/{total_sets}").format(
+            current_set=current_set,
+            total_sets=total_sets
+        )
```
There is not enough descriptive context in this TRANSLATOR_NOTE. This is how it is currently extracted for translation:

```
#. "current_set" = current set number (1-based)
#. "total_sets" = total number of sets
#: src/seedsigner/gui/screens/tools_screens.py
msgid "Word {current_set}/{total_sets}"
msgstr ""
```

Also the translator probably(?) doesn't need to know that the numbering is 1-based.

kdmukai commented on src/seedsigner/gui/screens/tools_screens.py:
```diff
@@ -198,6 +198,92 @@ def update_title(self) -> bool:
+        super().__post_init__()
+
+        self.components.append(TextArea(
+            text=_("(H)eads: 1   /   (T)ails: 0"),
```
Should avoid using space chars for formatting / alignment. We do it in a few places elsewhere, but it's not ideal.

Also note the presentation of how the 1/0 is explained in `ToolsCoinFlipEntryScreen`. It's probably worth following that approach.

And, actually... I don't know how messy this might get, but I'd love to see a cleanup + refactor to combine everything into just `ToolsCoinFlipEntryScreen`, but using the improved UX here (the larger keyboard buttons, my proposed "Heads" / "Tails" keyboard labels).

All three coin flip interactions (all in one go, setwise 11-bit entry, final 3-7 bits of entropy) are basically the same with only minor messaging differences that should be easy to handle.

kdmukai commented on src/seedsigner/gui/screens/tools_screens.py:
```diff
@@ -198,6 +198,92 @@ def update_title(self) -> bool:
+class ToolsCoinEntropySetwiseEntryScreen(KeyboardScreen):
+    current_set: int = 1
+    total_sets: int = 12
+    required_bits: int = 11
```
~~This is unnecessary. It's always going to be 11 bits.~~

kdmukai replied:
Ah, my mistake. Now that I'm reading through the View code, I get it.

And explanatory comment on this var wouldn't hurt.

kdmukai commented on src/seedsigner/gui/screens/tools_screens.py:
```diff
@@ -198,6 +198,92 @@ def update_title(self) -> bool:
+            screen_y=self.keyboard.rect[3] + 3*GUIConstants.COMPONENT_PADDING,
+        ))
+        self.components.append(TextArea(
+            text=_("Bits needed: {}").format(self.required_bits),
```
See my comment above about the var.

Also I wouldn't use the word "Bits" when we're asking the user to do coin flips.

kdmukai commented on src/seedsigner/gui/screens/tools_screens.py:
```diff
@@ -230,7 +316,7 @@ def __post_init__(self):
         self.rows = 1
         self.cols = 4
-        self.key_height = GUIConstants.get_top_nav_title_font_size() + 2 + 2*GUIConstants.EDGE_PADDING
+        self.key_height = GUIConstants.get_top_nav_title_font_size() + GUIConstants.EDGE_PADDING//4 + 2*GUIConstants.EDGE_PADDING
```
For these large keyboard keys, I think it makes sense to simplify all this and just the constant that's already defined for button height.

kdmukai commented on src/seedsigner/gui/screens/tools_screens.py:
```diff
@@ -198,6 +198,92 @@ def update_title(self) -> bool:
+        self.return_after_n_chars = self.required_bits
+        self.rows = 2
+        self.cols = 2
+        self.key_height = GUIConstants.get_top_nav_title_font_size() + 2*GUIConstants.EDGE_PADDING
```
See comment below on keyboard key height.

kdmukai commented on src/seedsigner/helpers/mnemonic_generation.py:
```diff
@@ -121,3 +135,10 @@ def generate_mnemonic_from_image(image, wordlist_language_code: str = SettingsCo
+
+
+
+def get_bip39_word(bits: str) -> str:
+    """ Convert 11-bit string to BIP39 word """
+    index = int(bits, 2)
+    return Seed.get_wordlist()[index]
```
This is never used.

kdmukai commented on src/seedsigner/helpers/mnemonic_generation.py:
```diff
@@ -90,14 +92,26 @@ def generate_mnemonic_from_coin_flips(coin_flips: str, wordlist_language_code: s
         * binary digit stream is treated as string data.
         * hashed via SHA256.
```
This is no longer true after this implementation was changed.

kdmukai commented on tests/test_mnemonic_generation.py:
```diff
@@ -195,3 +218,95 @@ def test_50_dice_rolls():
+    mnemonic = mnemonic_generation.generate_mnemonic_from_coin_flips(coin_flips)
+    actual = " ".join(mnemonic)
+    assert bip39.mnemonic_is_valid(actual)
+    assert actual == expected
```
I would refactor all of these tests into one list of tuples or dicts and just iterate through the test vectors to avoid all the duplicated test code.

The 256 bit and 128 bit tests can all be combined together.

kdmukai commented on tests/test_mnemonic_generation.py:
```diff
@@ -195,3 +218,95 @@ def test_50_dice_rolls():
+    with pytest.raises(ValueError) as e:
+        invalid_coin_flips = "01012" + "0" * 123  # 128 chars with a '2'
+        mnemonic_generation.generate_mnemonic_from_coin_flips(invalid_coin_flips)
+    assert "only contain '0' or '1'" in str(e)
```
This sort of assert might be a bit fragile, though any future refactoring of the error message would trigger an obvious failure here so it would be quickly caught.

I don't have a suggestion for an improvement. Just noting that this style of test always makes me uncomfortable, even when I do it myself.

kdmukai commented on tests/test_mnemonic_generation.py:
```diff
@@ -195,3 +218,95 @@ def test_50_dice_rolls():
+        mnemonic_generation.generate_mnemonic_from_coin_flips(invalid_coin_flips)
+    assert "128 or 256 bits" in str(e)
+
+    with pytest.raises(ValueError) as e:
+        invalid_coin_flips = "0" * 255  # Too short for 256
+        mnemonic_generation.generate_mnemonic_from_coin_flips(invalid_coin_flips)
+    assert "128 or 256 bits" in str(e)
```
I would refactor all of these into another list of tuples/dicts to streamline the test code.

Somewhere around 3 repetitions should be the point where you consider refactoring to be more DRY.

kdmukai commented on src/seedsigner/views/tools_views.py:
```diff
@@ -256,6 +264,194 @@ def run(self):
+        else:
+            # Final set collected; generate the full mnemonic
+            full_entropy = self.bits_collected[:self.total_flips]
+            entropy_bytes = int(full_entropy, 2).to_bytes(16 if self.total_flips == 128 else 32, byteorder='big')
+            mnemonic = mnemonic_generation.generate_mnemonic_from_bytes(entropy_bytes, wordlist_language_code=wordlist_language_code)
```
This should be using the `generate_mnemonic_from_coin_flips` utility function.

Even for straightforward operations we have to be able to point to one and only one place that is doing vital calculations that has unit test coverage.

kdmukai commented on src/seedsigner/views/tools_views.py:
```diff
@@ -256,6 +264,194 @@ def run(self):
+
+        if self.current_set <= self.total_sets:
+            # Convert the 11-bit set to a BIP-39 word
+            word_index = int(ret, 2)
+            word = Seed.get_wordlist(wordlist_language_code)[word_index]
```
This should be using the (currently unused) `get_bip39_word` method. And then that method should have some unit tests.

kdmukai commented on src/seedsigner/views/tools_views.py:
```diff
@@ -256,6 +264,194 @@ def run(self):
+        self.last_set_bits = 7 if total_flips == 128 else 3
+
+    def run(self):
+        wordlist_language_code = self.settings.get_value(SettingsConstants.SETTING__WORDLIST_LANGUAGE)
```
This should be moved further down to just before it's needed in the if/else.

kdmukai commented on src/seedsigner/gui/screens/tools_screens.py:
```diff
@@ -198,6 +198,92 @@ def update_title(self) -> bool:
+            current_set=current_set,
+            total_sets=total_sets
+        )
+        self.return_after_n_chars = self.required_bits
```
I'm a bit torn here. On the one hand I think it's a little silly to pass in `required_bits` when it serves the exact same purpose as `return_after_n_chars`.

But on the other hand, the `required_bits` var name does help a bit to explain the context of what's going on and what's required in this particular coin flipping use case.

I think I slightly lean towards keeping `required_bits` (though that name could be improved... `num_flips_required`, `num_flips_needed`???) but basically kind of dislike either option so there's no ideal outcome here.

kdmukai commented:
<img width="240" height="240" alt="ToolsMenuView" src="https://github.com/user-attachments/assets/25995255-8073-4f35-bb71-c7a716d31bff" />

Minor nit: I don't love the new coins icon here. Stacks of coins do not immediately make me think of coin flips. Also as an old tech guy, it looks like a database cluster to me.

Would @easyuxd want to take a crack at an icon for this?

kdmukai commented on src/seedsigner/views/tools_views.py:
```diff
@@ -256,6 +264,194 @@ def run(self):
+
+    def run(self):
+        self.run_screen(
+            ButtonListScreen,
```
I think I'd prefer a custom Screen that can display the bits the user entered that yielded this word.

kdmukai replied:
Also the word itself should be displayed as plain text, not rendered as a button.

The button should just say "Next".

kdmukai commented on src/seedsigner/views/tools_views.py:
```diff
@@ -256,6 +264,194 @@ def run(self):
+    def run(self):
+        self.run_screen(
+            ButtonListScreen,
+            title=_("word {}").format(self.current_set),
```
TopNav titles should always be in title case (first letter of each word capitalized).

Also this needs a TRANSLATOR_NOTE to provide context for the variable being inserted.

kdmukai commented on src/seedsigner/views/tools_views.py:
```diff
@@ -256,6 +264,194 @@ def run(self):
+
+class ToolsCoinEntropySetwiseBip39WordView(View):
+    """ Displays the BIP-39 word for the current set of coin flips. """
+    def __init__(self, total_flips: int, current_set: int, bits_collected: str, mnemonic: list, word: str):
```
Unfortunately I think `bits_collected` and especially `mnemonic` are unacceptable as View args.

In an official release image we write any console output to `/dev/null` but we still avoid ever writing any user secrets to the console just as a matter of principle.

With your usage here, we get:
```
2025-06-15 00:28:08,946     INFO [seedsigner.controller start (358)]: Appending next destination: ToolsCoinEntropySetwiseEntryView({'total_flips': 128, 'current_set': 3, 'bits_collected': '1010110101011110010110', 'mnemonic': ['pull', 'version']})
2025-06-15 00:28:08,952     INFO [seedsigner.controller start (308)]: 
back_stack: [
     7: ToolsCoinEntropySetwiseEntryView({'total_flips': 128, 'current_set': 3, 'bits_collected': '1010110101011110010110', 'mnemonic': ['pull', 'version']})
     6: ToolsCoinEntropySetwiseBip39WordView({'total_flips': 128, 'current_set': 2, 'bits_collected': '1010110101011110010110', 'mnemonic': ['pull', 'version'], 'word': 'version'})
     5: ToolsCoinEntropySetwiseEntryView({'total_flips': 128, 'current_set': 2, 'bits_collected': '10101101010', 'mnemonic': ['pull', 'version']})
     4: ToolsCoinEntropySetwiseBip39WordView({'total_flips': 128, 'current_set': 1, 'bits_collected': '10101101010', 'mnemonic': ['pull', 'version'], 'word': 'pull'})
     3: ToolsCoinEntropySetwiseEntryView({'total_flips': 128})
     2: ToolsCoinInputMethodView({'total_flips': 128})
     1: ToolsCoinEntropyMnemonicLengthView()
     0: ToolsMenuView()
]
2025-06-15 00:28:08,959     INFO [seedsigner.controller start (312)]: Executing ToolsCoinEntropySetwiseEntryView({'total_flips': 128, 'current_set': 3, 'bits_collected': '1010110101011110010110', 'mnemonic': ['pull', 'version']})
```

Instead you would have to add a `Controller` working var (e.g. `sign_message_data`). Or modify `Controller.storage` to hold a not-yet-complete pending mnemonic.

kdmukai replied:
Actually, see `SeedStorage.update_pending_mnemonic`. I think it can already do everything you need. And I see no need to keep around the previous coin flips once you have the next mnemonic word (I guess some reverse decoding from word back to bits would be necessary if the user hits the BACK button).

kdmukai commented on src/seedsigner/views/tools_views.py:
```diff
@@ -256,6 +264,194 @@ def run(self):
+
+    def run(self):
+        wordlist_language_code = self.settings.get_value(SettingsConstants.SETTING__WORDLIST_LANGUAGE)
+        required_bits = self.last_set_bits if self.current_set == self.total_sets + 1 else 11
```
I think a light refactor of `ToolsCalcFinalWordFinalizePromptView` would be much better here. That View could be modified to force the user to use coin flips to complete the final word (remove the other two options, replace "Coin flip entropy" with just "Next") and then we retain the explanation and display of how the final word was calculated.

It also cleans up the slight awkwardness that this flow shows every resulting word but does NOT show the final word.

The downside would be that "calc final word" flow jumps straight to finalizing the seed, whereas we would still want to review all the words one last time. Should be able to alter the flows to do that, but might get a bit complicated.

easyuxd commented:
> <img alt="ToolsMenuView" width="240" height="240" src="https://private-user-images.githubusercontent.com/934746/482236389-25995255-8073-4f35-bb71-c7a716d31bff.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTYyNDAzMTgsIm5iZiI6MTc1NjI0MDAxOCwicGF0aCI6Ii85MzQ3NDYvNDgyMjM2Mzg5LTI1OTk1MjU1LTgwNzMtNGYzNS1iYjcxLWM3YTcxNmQzMWJmZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgyNlQyMDI2NThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MGZlNGRkN2NiNDJmNDMyMjE3MGQ4ZjQwODAxOWYxNzE3OTNjOTMzYzJlZjY0MGQxMzVkOTA2NWE3M2RmYzI1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.oDI6pwMZt4utNEQcItDB64n-zse0O5SYTJ0cSiaNenU">
> Minor nit: I don't love the new coins icon here. Stacks of coins do not immediately make me think of coin flips. Also as an old tech guy, it looks like a database cluster to me.
> 
> Would @easyuxd want to take a crack at an icon for this?

@kdmukai 
Agreed -- this icon should be replaced. Unfortunately, this is the only free coin that FontAwesome allows for the time being. I have this on my list of outstanding icons that will finally replace the FontAwesome placeholders:

<img width="64" height="64" alt="coin" src="https://github.com/user-attachments/assets/ad73175d-6218-4d26-bea9-ea3cdd6af0b9" />

kdmukai commented:
I don't view the icon as a blocker for this PR and we can always update it whenever your icon is ready @easyuxd. No rush as far as I'm concerned.

Stargazer10101 commented:
Thanks for the code review. I'll surely work on all of your remarks.

Stargazer10101 committed:
address PR feedback

Stargazer10101 committed:
update submodule pointer to latest commit

Stargazer10101 committed:
address PR feedback

Stargazer10101 committed:
force coin flips for the final word in ToolsCalcFinalWordFinalizePromptView

Stargazer10101 committed:
add review words feature in calc final word flow

Stargazer10101 committed:
Merge branch 'dev' into create_seed_with_coin

notTanveer commented on src/seedsigner/views/tools_views.py:
```diff
@@ -23,11 +27,13 @@ class ToolsMenuView(View):
     KEYBOARD = ButtonOption("Calc 12th/24th word", FontAwesomeIconConstants.KEYBOARD)
-    ADDRESS_EXPLORER = ButtonOption("Address explorer")
-    VERIFY_ADDRESS = ButtonOption("Verify address")
+    ADDRESS_EXPLORER = ButtonOption("Address Explorer")
+    VERIFY_ADDRESS = ButtonOption("Verify Address")
```
use sentence case capitalization.

```suggestion
    ADDRESS_EXPLORER = ButtonOption("Address explorer")
    VERIFY_ADDRESS = ButtonOption("Verify address")
```

notTanveer commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1175,8 +1175,10 @@ def __init__(self, seed_num: int, num_words: int):
 
     def run(self):
         # TODO: Change this later to use the generic Screen input keyboard
+
```
why these unnecessary blank spaces here? 
try to keep the changes minimal.

notTanveer commented on src/seedsigner/views/tools_views.py:
```diff
@@ -260,6 +269,202 @@ def run(self):
+class ToolsCoinEntropyMnemonicLengthView(View):
+    def run(self):
+        # TRANSLATOR_NOTE: Inserts the number of coin flips needed for a 12-word mnemonic
+        twelve = _("12 words ({} flips)").format(mnemonic_generation.COIN__NUM_FLIPS__12WORD)
+        TWELVE = ButtonOption(twelve, return_data=mnemonic_generation.COIN__NUM_FLIPS__12WORD)
+
+        # TRANSLATOR_NOTE: Inserts the number of coin flips needed for a 24-word mnemonic
+        twenty_four = _("24 words ({} flips)").format(mnemonic_generation.COIN__NUM_FLIPS__24WORD)
+        TWENTY_FOUR = ButtonOption(twenty_four, return_data=mnemonic_generation.COIN__NUM_FLIPS__24WORD)
```
i guess this can be done in two lines, with the help of a ternary operator.
this reduces the technical debt.
(although it's fine to keep this as is, this is just my take)

notTanveer commented on tests/test_flows_seed.py:
```diff
@@ -449,25 +449,33 @@ def load_right_seed_into_decoder(view: View):
             FlowStep(seed_views.SeedTranscribeSeedQRFormatView, button_data_selection=seed_views.SeedTranscribeSeedQRFormatView.STANDARD_12),
             FlowStep(seed_views.SeedTranscribeSeedQRWarningView),
             FlowStep(seed_views.SeedTranscribeSeedQRWholeQRView),
+
```
why  the blank spaces here?

Stargazer10101 committed:
use sentence case capitalization and remove blank lines

Stargazer10101 committed:
fix: correct BIP39 word expectations in mnemonic generation tests

Stargazer10101 committed:
fix screenshot generator constructor parameters for ToolsCoinEntropySetwiseBip39WordView

Stargazer10101 commented:
Hi @kdmukai Can you please review the PR ?
